### PR TITLE
Added Support for App & No Authentication

### PR DESCRIPTION
**Pull request by ilamp**
- This adds support for For app & no authentication.
- Added/Updated PHPDocs

### CODE REVIEW

**Comment by ilamp** [comment]
@freekmurze Will you be able to review this? If you require any changes I can go ahead and make those. Thank you!

File: src/Client.php [hunk 1]
```diff
@@ -16,41 +16,63 @@
...
      * @param int $maxChunkSize Set max chunk size per request (determines when to switch from "one shot upload" to upload session and defines chunk size for uploads via session).
      * @param int $maxUploadChunkRetries How many times to retry an upload session start or append after RequestException.
      */
-    public function __construct(string $accessToken, GuzzleClient $client = null, int $maxChunkSize = self::MAX_CHUNK_SIZE, int $maxUploadChunkRetries = 0)
+    public function __construct($accessTokenOrAppCredentials = null, GuzzleClient $client = null, int $maxChunkSize = self::MAX_CHUNK_SIZE, int $maxUploadChunkRetries = 0)
     {
-        $this->accessToken = $accessToken;
+        if (is_array($accessTokenOrAppCredentials)) {
+            [$this->appKey, $this->appSecret] = $accessTokenOrAppCredentials;
+        } elseif (is_string($accessTokenOrAppCredentials)) {
```
Comment (freekmurze) [inline src/Client.php]: I really don't like `else` statements. Use an second `if` statement.

File: src/Client.php [hunk 2]
```diff
@@ -661,11 +762,23 @@ public function setAccessToken(string $accessToken): self
...
     {
-        return array_merge([
-            'Authorization' => "Bearer {$this->accessToken}",
-        ], $headers);
+        $auth = []; // defaults no-auth, can be used for public endpoints
+        if (isset($this->appKey, $this->appSecret)) {
+            $auth = [
+                'Authorization' => 'Basic '.base64_encode("{$this->appKey}:{$this->appSecret}"),
+            ];
+        } elseif (isset($this->accessToken)) {
```
Comment (freekmurze) [inline src/Client.php]: `else` is the bane of my existence. Refactor this in dedecated functions `getHeadersForCredentials` and `getHeadersForBearerToken`. Use a ternary for determining which of the functions you need.

**Comment by ilamp** [comment]
Thank you @freekmurze. Requested changes are in.

FYI: I'm also considering adding few more exception types. Please let me know if that's okay.

**Comment by ilamp** [comment]
@freekmurze Will you be able to merge my pull request sometime soon? Please let me know if I can do anything to help.

File: src/Client.php [hunk 3]
```diff
@@ -16,41 +16,64 @@
 
 class Client
 {
-    const THUMBNAIL_FORMAT_JPEG = 'jpeg';
-    const THUMBNAIL_FORMAT_PNG = 'png';
+    public const THUMBNAIL_FORMAT_JPEG = 'jpeg';
```
Comment (freekmurze) [inline src/Client.php:19]: Please don't add `public` here.

File: src/Client.php [hunk 4]
```diff
@@ -16,41 +16,64 @@
...
-    const MAX_CHUNK_SIZE = 1024 * 1024 * 150;
+    public const MAX_CHUNK_SIZE = 1024 * 1024 * 150;
 
-    const UPLOAD_SESSION_START = 0;
-    const UPLOAD_SESSION_APPEND = 1;
+    public const UPLOAD_SESSION_START = 0;
+    public const UPLOAD_SESSION_APPEND = 1;
 
-    /** @var string */
+    /**
```
Comment (freekmurze) [inline src/Client.php:33]: Leave the formatting as it is.

File: src/Client.php [hunk 5]
```diff
@@ -64,6 +87,11 @@ public function __construct(string $accessToken, GuzzleClient $client = null, in
      * If the source path is a folder all its contents will be copied.
      *
      * @link https://www.dropbox.com/developers/documentation/http/documentation#files-copy_v2
+     *
+     * @param string $fromPath
```
Comment (freekmurze) [inline src/Client.php:91]: Do not add docs block where everything is already type hinted.

File: src/Client.php [hunk 6]
```diff
@@ -64,6 +87,11 @@ public function __construct(string $accessToken, GuzzleClient $client = null, in
...
+     * @param string $toPath
+     * @return array
+     * @throws Exception
```
Comment (freekmurze) [inline src/Client.php:94]: Do not add exception docblocks

File: src/Client.php [hunk 7]
```diff
@@ -683,11 +785,37 @@ public function setAccessToken(string $accessToken): self
...
     /**
      * Get the HTTP headers.
+     *
+     * @param array $headers
+     * @return array
      */
     protected function getHeaders(array $headers = []): array
     {
-        return array_merge([
+        $auth = []; // defaults no-auth, can be used for public endpoints
```
Comment (freekmurze) [inline src/Client.php:794]: Remove the comment

Comment (ilamp) [reply src/Client.php]: Removed the comment.

File: src/Client.php (hunk 3 above)
Comment (ilamp) [reply src/Client.php]: Removed the access modifiers. I can open a new pull request for this later

File: src/Client.php (hunk 4 above)
Comment (ilamp) [reply src/Client.php]: Reverting changes

File: src/Client.php (hunk 6 above)
Comment (ilamp) [reply src/Client.php]: Removed

File: src/Client.php (hunk 5 above)
Comment (ilamp) [reply src/Client.php]: Removed

**Comment by ilamp** [comment]
- Added a couple of quick tests
- Removed the unrelated changes

Thank you for reviewing the pull request!

> Thanks for your work on this.
> 
> This PR contains some unwanted changes. Only PR the bare necessities to make the new functionality work. Could you also add tests?

**Comment by freekmurze** [comment]
Thanks!

**Pull request merged**